list.c

```diff
@@ -184,15 +184,24 @@ void vListInsert( List_t * const pxList,
         *   4) Using a queue or semaphore before it has been initialised or
         *      before the scheduler has been started (are interrupts firing
         *      before vTaskStartScheduler() has been called?).
-        *   5) If the FreeRTOS port supports interrupt nesting then ensure that
+        *   5) Attempting to 'take' binary semaphores created using
+        *      `xSemaphoreCreateBinary()` or `xSemaphoreCreateBinaryStatic()`
+        *      APIs, before 'giving' them. Binary semaphores created using
+        *      `xSemaphoreCreateBinary()` or `xSemaphoreCreateBinaryStatic()`,
+        *      are created in a state such that the semaphore must first be
+        *      'given' using xSemaphoreGive() API before it can be 'taken' using
+        *      xSemaphoreTake() API.
```

Comment:
This new statement seems misleading.  Taking a binary semaphore that has not yet been given results in a yield, not getting stuck in this loop.  However, taking a binary semaphore before it has been _created_ does end up this way.  Item #4 above covers that case already.  

There's nothing wrong with attempting to take a binary semaphore that has not yet been given -- in fact that is the most common use of a binary semaphore -- so maybe we're better off without this new statement in the code.

Reply:
You are exactly right @jefftenney. I verified that with the following code on a Cortex-M7:

```c
void Task( void * param )
{
    SemaphoreHandle_t xBinarySem;

    ( void ) param;

    xBinarySem = xSemaphoreCreateBinary();

    xSemaphoreTake( xBinarySem, portMAX_DELAY );

    for( ;; )
    {
        HAL_GPIO_TogglePin( LD1_GPIO_Port, LD1_Pin );
        vTaskDelay( pdMS_TO_TICKS( 1000 ) );
    }
}
void app_main( void )
{
    xTaskCreate( Task,
                 "Task",
                 configMINIMAL_STACK_SIZE,
                 NULL,
                 tskIDLE_PRIORITY | portPRIVILEGE_BIT,
                 NULL );

    vTaskStartScheduler();

    for( ;; );
}
```

I'll revert it in a separate PR.

Reply:
Removed in this PR - https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/1056. Thanks again @jefftenney !

Reply:
@jefftenney and @aggarg , thank you both for taking a look at this. Is there an explanation for the behavior I was seeing, then? I have a `TaskSetup()` function inside `main()` which gets called before we call `vTaskStartScheduler();` and the end of `main()`. `TaskSetup()` contains, among other things, this: 

```c
xI2C1mutex = xSemaphoreCreateBinaryStatic(&xI2C1mutexBuffer);	// I2C1 bus
xI2C2mutex = xSemaphoreCreateBinaryStatic(&xI2C2mutexBuffer);	// I2C2 bus
xI2C4mutex = xSemaphoreCreateBinaryStatic(&xI2C4mutexBuffer);	// I2C4 bus
```

Like that, my debugger gets stuck forever inside of `vListInsert()` in `list.c`. 

If I change it to this, however, it works fine: 
```c
xI2C1mutex = xSemaphoreCreateBinaryStatic(&xI2C1mutexBuffer);	// I2C1 bus
xI2C2mutex = xSemaphoreCreateBinaryStatic(&xI2C2mutexBuffer);	// I2C2 bus
xI2C4mutex = xSemaphoreCreateBinaryStatic(&xI2C4mutexBuffer);	// I2C4 bus

xSemaphoreGive(xI2C1mutex);
xSemaphoreGive(xI2C2mutex);
xSemaphoreGive(xI2C4mutex);
```

Furthermore, I think bullet 4 still needs something extra. It says: 

```
*   4) Using a queue or semaphore before it has been initialised or
*      before the scheduler has been started (are interrupts firing
*      before vTaskStartScheduler() has been called?).
```

So, does this include regular mutexes, recursive mutexes, *and* binary semaphores? Or, just binary semaphores? I mean, a mutex is a type of semaphore. 

Also, what does it mean to "be initialised"? 

@jefftenney , your wording above, "taking a binary semaphore before it has been _created_", makes more sense than using the word "initialised", _unless_ initialized means something different, such as "created *and* taken", as seems to be implied by the [official documentation](https://freertos.org/xSemaphoreCreateBinaryStatic.html):

> The semaphore is created in the 'empty' [I read this as "uninitialized"] state, meaning the semaphore must first be given using the xSemaphoreGive() API function before it can subsequently be taken (obtained) using the xSemaphoreTake() function.

So, _something_ in this documentation needs to be improved and or clarified.

Reply:
@ElectricRCAircraftGuy The term "create" and "initialize" are equivalent here.  One possible improvement to the official docs would be to use "create" exclusively, because that term matches the API function names.

The term "empty" does not refer to the created/uncreated state.  Instead it describes one of the two states of a semaphore after it has been created.  It is either empty or full.  Some people say taken or given.  Some people say unsignaled or signaled.  Application usage of the semaphore changes it back and forth between these two states.

Another improvement could be to add clarity as you suggested.  Something like this (probably needs work):

_The semaphore is created in the 'empty' state, meaning the semaphore must first be given using the xSemaphoreGive() API function before xSemaphoreTake() can successfully take the semaphore._

All of this still leaves you with your mystery.  Do any of the 5 items listed in the code comment apply?  Something might be corrupting kernel data structures and your startup calls to xSemaphoreGive() might be hiding the issue.